accepted/2025/dotnet-list-package-restore-no-op-check.md

```diff
@@ -0,0 +1,107 @@
+However, it is possible a user might have run restore once and followed it up with a modification to the project file.
+That means the assets file has information that has expired.
+This is not detected/asserted by the `dotnet list package command`.
+As a result, in this specific scenario, `dotnet list package` command misinforms users about their projects.
+
+## Explanation
+
+### Functional explanation
+
+Assuming the functionality has been implemented, here is what the expected interaction with a user would look like.
+
+#### **Scenario 1: User runs `dotnet restore` and modifies the project file**
+
+*User Action:*
+
+- Runs `dotnet restore` to restore packages for their project.
+- Modifies the project file by adding or removing a package reference.
+- Runs `dotnet list package`.
+
+*Expected Behavior:*
+
+- The command checks if the assets file is in sync with the project file.
+- If the assets file is outdated, it prompts the user to run `dotnet restore` again.
+
+#### **Scenario 2: User runs `dotnet restore` and does not modify the project file**
+
+*User Action:*
+
+- Runs `dotnet restore`.
+- Runs `dotnet list package`.
+
+*Expected Behavior:*
+
+- The command lists packages accurately without any prompts.
+
+### Technical explanation
+
+<!-- Explain the proposal in sufficient detail with implementation details, interaction models, and clarification of corner cases. -->
+
+Comparing the contents of the project file and the assets file is the most important step here. We will achieve this by using no-op restore.
+
+- Create a DGSpec file from the evaluated project
+  - To create the DGSpec file, use an MSBuild target, [GenerateRestoreGraphFile](https://github.com/NuGet/NuGet.Client/blob/cb0b5561f4fd98347cedac11919f769da0b767f4/src/NuGet.Core/NuGet.Build.Tasks/NuGet.RestoreEx.targets#L46), that writes the DGSpec to a file and then load it.
+- Hash the DGSpec
+- Compare the hash value with the one saved in `project.nuget.cache` file
+
+#### If unable to load DGSpec file
+
+Log the following [similar to the dotnet sdk](https://github.com/dotnet/sdk/blob/b324c2e074f7691705b85b6ed69be23d350164b8/src/Cli/dotnet/commands/dotnet-add/dotnet-add-package/LocalizableStrings.resx#L147)
+
+> Error: Unable to create dependency graph file for project '{0}'.
+
+#### If the hash values are different
+
+Log the following
+
+> Error: The project's dependency graph has been altered. Please perform a restore for project '{0}'.
+
+## Drawbacks
+
+<!-- Why should we not do this? -->
+- Performance Overhead: Generating and hashing the DGSpec file adds slight overhead to the dotnet list package.
+
+## Rationale and alternatives
+
+<!-- Why is this the best design compared to other designs? -->
+<!-- What other designs have been considered and why weren't they chosen? -->
+<!-- What is the impact of not doing this? -->
+- **Force Restore Every Time**: Automatically run restore when dotnet list package is called.
```

Comment:
I think this should say "Implicitly run restore which might do a no-op with an option to disable via a command-line argument `--no-restore` which other commands have.  This would probably be me vote personally.

Reply:
https://github.com/nuget/home/issues/13406 is the tracking issue.

It'd be worth to call out the challenges if we were to go that direction.

We can also see if @OliaG can help with some CD to see if the breaking is worth it.

Reply:
fwiw, as per the spec guide, I think we should make a good argument about why we're not pursuing certain alternative. 
In this case, those drawbacks are not obvious.